KAFKA-4451: Fix OffsetIndex overflow when replicating a highly compacted topic. OffsetIndex provides checks for overflow, used when deciding to roll a LogSegment

michaelschiff opened the pull request:
Fix OffsetIndex overflow when replicating a highly compacted topic.
https://issues.apache.org/jira/browse/KAFKA-4451

michaelschiff committed:
Adds a unit test case to LogTest that currently fails due to the
presence of KAFKA-4451

michaelschiff commented:
so far this just adds a unit test that fails due to the bug - real fix coming next.

michaelschiff committed:
adds additional requirement to OffsetIndex.append - checks that offset to add is not more than an integer above the base offset

michaelschiff committed:
OffsetIndex.canAppend(Long) - checks to see if the index is not full and
this offset is not more than an integer above the base
Log.maybeRoll - accepts a new parameter, minOffsetInMessages, which is
passes to the offsetIndex.canAppend to check if this message batch can
be safely added to the current index.

michaelschiff committed:
OffsetIndex.canAppend doesnt check for index full, LogTest closes and deletes all resources that it opens

ijuma commented:
cc @becketqin @junrao

ijuma commented:
Thanks for the PR, can you please update the PR title to be a bit more informative? For example, `KAFKA-4451: Fix OffsetIndex overflow when replicating highly compacted topic`. The PR description can then be edited to be blank as it doesn't add any useful information (this matters because the PR title becomes the commit title and the PR description becomes the rest of the commit message in the squashed commit that is merged).

michaelschiff commented:
Got it. Let me know if anything else looks strange to you.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -407,7 +407,8 @@ class Log(@volatile var dir: File,
         val segment = maybeRoll(messagesSize = validMessages.sizeInBytes,
-                                maxTimestampInMessages = appendInfo.maxTimestamp)
+                                maxTimestampInMessages = appendInfo.maxTimestamp,
+                                minOffsetInMessages = appendInfo.firstOffset)
```
It seems that we should be passing in appendInfo.lastOffset instead of firstOffset. The reason is that when appending to the time index, we use the offset of the message with the largest timestamp. That offset could be as large as the last offset.

michaelschiff replied:
You're absolutely correct, this should use lastOffset.

junrao commented on core/src/main/scala/kafka/log/OffsetIndex.scala:
```diff
@@ -114,13 +114,22 @@ class OffsetIndex(file: File, baseOffset: Long, maxIndexSize: Int = -1)
+    * @param offset that we want to potentially append
+    * @return true if it is safe to append this offset to this index
+    */
+  def canAppend(offset: Long) : Boolean = {
```
It seems that this method should be in Log, instead of here?

michaelschiff replied:
It could go there, but putting it in Log seems to imply that it is the Log that can(not) be appended to, however what it checks is if the LogSegment (or really the OffsetIndex) will be overflow.  Maybe it belongs in LogSegment?

ijuma replied:
One more thing: having to call both `isFull` and `canAppend` methods in the `OffsetIndex` is not ideal. If we decide that a `canAppend` makes sense here, then it would probably be better for it to call `isFull` as well. In any case, let's hear Jun's reasoning for his suggestion first.

junrao replied:
My reasoning was that if we put the check here, we have to duplicate the same check in TimeIndex as well. So, perhaps we could add this method and the assertion in LogSegment instead.

Also, canAppend() is a bit too general. Perhaps we could rename it to sth like canConvertToRelativeOffset(offset)?

michaelschiff replied:
Makes sense.  I will rename the method and relocate the method and assertion

junrao reviewed (commented):
Thanks for the patch. Great finding! Left a couple of comments.

michaelschiff committed:
maybeRoll takes maxOffsetInMessages not min

ijuma commented on core/src/main/scala/kafka/log/OffsetIndex.scala:
```diff
@@ -114,13 +114,22 @@ class OffsetIndex(file: File, baseOffset: Long, maxIndexSize: Int = -1)
   def append(offset: Long, position: Int) {
     inLock(lock) {
       require(!isFull, "Attempt to append to a full index (size = " + _entries + ").")
+      require(!((offset - baseOffset) > Integer.MAX_VALUE), "Attempting to pack too large an offset range into index.")
```
It would be good to avoid duplicating logic here, we should call the other method instead.

michaelschiff replied:
Agree, I will replace this with a call to the other method.

ijuma commented on core/src/test/scala/unit/kafka/log/LogTest.scala:
```diff
@@ -1013,6 +1013,61 @@ class LogTest extends JUnitSuite {
+    //This write will roll the segment, and create a new segment with baseOffset = log end of previous segment  1
+    log.append(set2, false)
+
+    //This write will also roll the segment, ""                                       "" Integer.MAX_VALUE + 2
```
Is the white space intentional?

ijuma commented on core/src/test/scala/unit/kafka/log/LogTest.scala:
```diff
@@ -1013,6 +1013,61 @@ class LogTest extends JUnitSuite {
+    log.append(set4, false)
+
+    log.close()
+    val indexFiles = logDir.listFiles(new FilenameFilter {
```
Nit: it's fine to simply do `logDir.listFiles.asScala.filter(..)` instead of passing a `FilenameFilter`.

ijuma commented on core/src/test/scala/unit/kafka/log/LogTest.scala:
```diff
@@ -1013,6 +1013,61 @@ class LogTest extends JUnitSuite {
+      override def accept(dir: File, name: String)  = name.contains(".index")
+    });
+    assertEquals(3, indexFiles.length)
+    for (file : File <- indexFiles) {
```
The `: File` is not needed here.

ijuma commented on core/src/test/scala/unit/kafka/log/LogTest.scala:
```diff
@@ -1013,6 +1013,61 @@ class LogTest extends JUnitSuite {
+    });
+    assertEquals(3, indexFiles.length)
+    for (file : File <- indexFiles) {
+      val offsetIndex = new OffsetIndex(file, file.getName.replace(".index","").toLong)
```
Nit: space missing after `,`.

ijuma reviewed (commented):
Thanks for the PR, left a few minor comments.

michaelschiff committed:
removed whitespace and unecessary type declaraction

michaelschiff committed:
remove duplicate logic from OffsetIndex.append

michaelschiff committed:
more idiomatic file filter in LogTest.testOverCompactedLogRecovery

junrao commented on core/src/test/scala/unit/kafka/log/LogTest.scala:
```diff
@@ -1013,6 +1013,54 @@ class LogTest extends JUnitSuite {
+      compressionCodec = NoCompressionCodec,
+      offsetSeq = Seq(0),
+      messages = new Message("v1".getBytes(), "k1".getBytes(), Message.NoTimestamp, magicValue = Message.CurrentMagicValue))
+    val set2 = new ByteBufferMessageSet(
```
This is a great example. Now, I am wondering if we have fixed the issue completely. What happens if we append two messages at offset 0 and at offset Integer.MAX_VALUE.toLong + 2 in succession? On appending the second message, we will first roll a new segment at base offset 1, but the appending of the second message will fail since the offset difference is larger than Integer.MAX_VALUE.

junrao replied:
To get around this, we could roll a segment with a base offset of the first offset in the to be appended message set, instead of log end offset. However, it's not easy to get the first offset in the follower w/o decompression since the offset of the wrapper message is always the offset of the last message in the set. KIP-98 proposes a new message format that allows us to get the first offset w/o decompression. Not sure if there are other workarounds in the interim.

michaelschiff replied:
I see what you're saying, good catch. 
This is probably worse than just decompressing the message, but another potential work-around would be to iterate the maybeRoll() logic. In this case (using your same example), we would try to append the message set, it would roll the log with a new base offset 1, check and see that it could still not append, roll another new log with base offset 2 which we are able to append to.  The cost of this approach would be bounded by the span between the two messages.

michaelschiff replied:
Your comment actually makes me aware of another issue:

It seems assumed that a single message set will fit into the same index (at worst it doesnt have room for the whole thing and we roll a new one). This assumption can not always be upheld in the case of highly compact topics. For example this message set:
`new ByteBufferMessageSet(
      compressionCodec = NoCompressionCodec,
      offsetSeq = Seq(0, Integer.MAX_VALUE.toLong + 2),
      messages = new Message(...),
      new Message(...))`

must be split across two log segments.

becketqin replied:
@michaelschiff Currently the offset within the same MessageSet would not differ by more than Integer.MAX_VALUE. Otherwise the message would already hit the max message size. During log compaction we do not merge MessageSet either. So maybe we don't need to worry too much about this.

If we assume the above is the true. As a short term fix, when log roll happens, we can always roll a new log segment with base offset `max(logEndOffset, largestOffsetOfTheMessageSetToAppend - Integer.MAX_VALUE)`. So in the case of what @junrao raised: two consecutive message offsets of 0 and `Integer.MAX_VALUE.toLong + 2`, the newly rolled out segment will have `max(1, Integer.MAX_VALUE.toLong + 2 - Integer.MAX_VALUE.toLong)` which is 2.

This is a little hacky but seems working.

michaelschiff replied:
@becketqin okay great, I didn't know that this guarantee would be upheld.  I will implement this solution then: Given that a MessageSet will never contain an offset range greater than an integer, largestOffsetOfTheMessageSetToAppend - Integer.MAX_VALUE is guaranteed to be <= lowestOffsetOfTheMessageSetToAppend, Allowing us to pick the max of this and the log end offset from last time (since subtracting the full Integer width might have moved us back farther than we needed.)

michaelschiff committed:
removed OffsetIndex.canAppend and the assertion in OffsetIndex.append.
added LogSegment.canConvertToRelativeOffset which is checked in Log.maybeRoll and asserted in LogSegment.append

michaelschiff committed:
updated ByteBufferMessageSet.FilterResult to include the maxOffset.
ByteBufferMessageSet.filterInto keeps track of this.  LogCleaner uses
the maxOffset from the FilterResult to pass to LogSegment.append.
Update LogSegment UTs to pass in the new parameter

michaelschiff committed:
supply base offset to Log.roll which chooses the max of this base offset and the log end offset from last time.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -744,7 +744,12 @@ class Log(@volatile var dir: File,
+      /*
+        maxOffsetInMessages - Integer.MAX_VALUE is a heuristic value for the first offset in the set of messages.
+        Since the offset in messages will not differ by more than Integer.MAX_VALUE, this is guaranteed <= the real
+        min offset
```
Bucket's suggestion seems to be the best option for now. Since it's a bit hachy, could we document a bit more why we need this, especially with respect to compacted topic?

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -756,10 +761,10 @@ class Log(@volatile var dir: File,
    * @return The newly rolled segment
    */
-  def roll(): LogSegment = {
+  def roll(offset: Long): LogSegment = {
```
Could we make offset defaults to 0 here? Also, would it be better to name offset sth like expectedNextOffset?

junrao commented on core/src/test/scala/unit/kafka/log/LogTest.scala:
```diff
@@ -1031,23 +1031,23 @@ class LogTest extends JUnitSuite {
+    //This write will roll the segment, yielding a new segment with base offset 2
     log.append(set2, false)
-    //This write will also roll the segment, ""                                       "" Integer.MAX_VALUE + 2
+    //This will also roll the segment, yielding a new segment with base offset Integer.MAX_VALUE+3
```
Those comments are really good. Could we actually verify the number of segments and the starting base offsets in the test?

michaelschiff replied:
Definitely. I will add assertions for the base offset.

michaelschiff committed:
default value 0 for Log.roll(Long)

michaelschiff committed:
better comments about the heuristic baseOffset value passed to Log.roll in Log.maybeRoll

michaelschiff committed:
added assertions and improved comments in LogTest.testOverCompactedLogRecovery

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -732,18 +733,26 @@ class Log(@volatile var dir: File,
+      /*
+        maxOffsetInMessages - Integer.MAX_VALUE is a heuristic value for the first offset in the set of messages.
+        Since the offset in messages will not differ by more than Integer.MAX_VALUE, this is guaranteed <= the real
+        min offset. Prior behavior assigned new baseOffset = logEndOffset from old segment.  This was problematic in
```
"min offset" => "first offset in the set. Determining the true first offset in the set requires decompression, which the follower is trying to avoid during log append."

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -649,7 +650,7 @@ class Log(@volatile var dir: File,
         // we must always have at least one segment, so if we are going to delete all the segments, create a new one first
         if (segments.size == numToDelete)
-          roll()
+          roll(0L)
```
This can just be roll().

junrao commented on core/src/main/scala/kafka/log/LogSegment.scala:
```diff
@@ -74,7 +74,14 @@ class LogSegment(val log: FileMessageSet,
 
-  /**
+   /**
+     * checks that the argument offset can be represented as an integer offset relative the baseOffset.
```
relative => relative to

junrao commented on core/src/main/scala/kafka/log/LogSegment.scala:
```diff
@@ -86,14 +93,15 @@ class LogSegment(val log: FileMessageSet,
    */
   @nonthreadsafe
-  def append(firstOffset: Long, largestTimestamp: Long, offsetOfLargestTimestamp: Long, messages: ByteBufferMessageSet) {
+  def append(firstOffset: Long, largestTimestamp: Long, offsetOfLargestTimestamp: Long, largestOffset: Long, messages: ByteBufferMessageSet) {
```
Could we move largestOffset to after firstOffset?

junrao commented on core/src/main/scala/kafka/message/ByteBufferMessageSet.scala:
```diff
@@ -470,6 +472,9 @@ class ByteBufferMessageSet(val buffer: ByteBuffer) extends MessageSet with Loggi
             offsetOfMaxTimestamp = shallowOffset
           }
+          if (shallowOffset > maxOffset) {
+            maxOffset = shallowOffset
```
It seems maxOffset needs to be computed in the else part too?

michaelschiff committed:
improved comments. changed parameter order of LogSegment.append. rely on default parameter value for Log.roll() where applicable

Commit:
Because messages have strictly increasing offsets ByteBufferMessageSet.filterInto updates maxOffset for every message iterated over

michaelschiff committed:
maintain maxOffset in ByteBufferMessageSet.filterInto deep iterator condition

michaelschiff committed:
Merge branch 'trunk' into bug/4451

junrao commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java:
```diff
@@ -95,7 +95,7 @@ public int validBytes() {
     public FilterResult filterTo(LogEntryFilter filter, ByteBuffer buffer) {
         long maxTimestamp = Record.NO_TIMESTAMP;
-        long shallowOffsetOfMaxTimestamp = -1L;
+        long maxOffset = -1L, shallowOffsetOfMaxTimestamp = -1L;
```
Could we split the two variables into two different lines?

junrao commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java:
```diff
@@ -142,6 +142,9 @@ public FilterResult filterTo(LogEntryFilter filter, ByteBuffer buffer) {
                     shallowOffsetOfMaxTimestamp = shallowEntry.offset();
                 }
+                if (shallowEntry.offset() > maxOffset) {
+                    maxOffset = shallowEntry.offset();
```
We need to maintain maxOffset in the else if case when writeOriginalEntry is false, right?

michaelschiff replied:
Yeah you're correct, I've fixed this now.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -453,7 +461,7 @@ class Log(@volatile var dir: File,
     var shallowMessageCount = 0
     var validBytesCount = 0
-    var firstOffset, lastOffset = -1L
+    var firstOffset, lastOffset, maxOffset = -1L
```
Could we split the two variables into two different lines?

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -465,6 +473,9 @@ class Log(@volatile var dir: File,
         monotonic = false
+      // keep track of the highest offset seen
+      if (maxOffset < entry.offset)
+        maxOffset = entry.offset
```
Hmm, do we need to maintain maxOffset? Isn't it the same as lastOffset?

michaelschiff replied:
I was under the impression that since the offsets may not be monotonically increasing, the lastOffset is not necessarily the maximum in the set.

junrao replied:
Yes, offsets are guaranteed to be monotonically increasing. If not, we throw an exception in Log.append. Timestamp, on the other hand, may not be monotonically increasing.

michaelschiff replied:
Got it. I thought because of the check on line 474/475 that it was accepted for offsets to come in out of order, but I see on line 405 if we are taking the offsets we are given, we check that the offsets are monotonic, and throw an exception otherwise. Ill remove this new field and book-keeping and just use lastOffset.

junrao commented on core/src/main/scala/kafka/log/LogSegment.scala:
```diff
@@ -91,17 +98,17 @@ class LogSegment(val log: FileRecords,
+  def append(firstOffset: Long, largestOffset: Long, largestTimestamp: Long, shallowOffsetOfMaxTimestamp: Long, records: MemoryRecords) {
     if (records.sizeInBytes > 0) {
-      trace("Inserting %d bytes at offset %d at position %d with largest timestamp %d at shallow offset %d"
+      trace("Inserting %d bytes at offset %d at position %d with largest timestamp %d at offset %d"
```
I think "shallow offset" is actually correct since we fixed the logic in KAFKA-4390.

junrao commented on core/src/test/scala/unit/kafka/log/LogCleanerTest.scala:
```diff
@@ -157,7 +157,7 @@ class LogCleanerTest extends JUnitSuite {
     log.append(record(0,0)) // offset 4
     // roll the segment, so we can clean the messages already appended
-    log.roll()
+    log.roll(5)
```
Are the changes in this file needed since by default log.roll() will just roll to log end offset?

michaelschiff replied:
my mistake, they are not needed

junrao commented on core/src/test/scala/unit/kafka/log/LogTest.scala:
```diff
@@ -1015,6 +1015,46 @@ class LogTest extends JUnitSuite {
+    //Writes into an empty log with baseOffset 0
+    log.append(set1, false)
+    assertEquals(0L, log.activeSegment.baseOffset)
+    //This write will roll the segment, yielding a new segment with base offset = max(1, 2) = 2
```
To be consistent with the pattern used in line 1040, max(1,2) should be max(2,1) since log end offset is the 2nd parameter?

junrao commented:
@michaelschiff : Thanks for rebasing the patch. Left a few more comments.

michaelschiff committed:
split variable declartion into multiple lines

michaelschiff committed:
removed unecessary argument from calls to Log.roll in LogCleanerTest

michaelschiff committed:
fixed trace log statement in LogSegment.append

michaelschiff committed:
MemoryRecords.filterInto maintains maxOffset when writeOriginalEntry is false too

michaelschiff committed:
consistent ordering to max(expected, logEnd) comment in LogTest

junrao commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java:
```diff
@@ -156,10 +160,13 @@ public FilterResult filterTo(LogEntryFilter filter, ByteBuffer buffer) {
                     maxTimestamp = info.maxTimestamp;
                     shallowOffsetOfMaxTimestamp = info.shallowOffsetOfMaxTimestamp;
                 }
+                if (shallowEntry.offset() > maxOffset) {
```
Hmm, we can't use shallowEntry.offset() since not all messages will be retained. Perhaps we could just maintain maxOffset as we add messages to retainedEntries.

junrao reviewed (commented):
Thanks for the updated patch. A couple of more comments.

michaelschiff committed:
use lastOffset instead of maxOffset since they are the same. dont track maxOffset

michaelschiff committed:
MemoryRecords.filterInto maintain maxOffset while iterating over records deciding what to retain

michaelschiff commented:
I am noticing the build seems to fail randomly (independent from the code changes). Can I re-trigger this without pushing new changes / is this something I should be concerned about?

ijuma commented:
retest this please

junrao commented:
@michaelschiff : Thanks for the patch. LGTM
